src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,763 @@
+      // within a single Vec<PathGlob> can result in having multiple different DirWildcard instances
+      // created after matching against a single directory node! So we do *not* mark
+      // `cur_dir_directories` as `mut`.
+      let mut cur_dir_files: IndexMap<PathBuf, remexec::FileNode> = cur_dir
+        .get_files()
+        .to_vec()
+        .into_iter()
+        .map(|file_node| (PathBuf::from(file_node.get_name()), file_node))
+        .collect();
+      let cur_dir_directories: IndexMap<PathBuf, remexec::DirectoryNode> = cur_dir
+        .get_directories()
+        .to_vec()
+        .into_iter()
+        .map(|directory_node| (PathBuf::from(directory_node.get_name()), directory_node))
+        .collect();
+
+      // 1b. Filter files and directories by globs.
+      let MultipleGlobs { include, exclude } = multiple_globs;
+      let mut globbed_files: IndexMap<PathBuf, remexec::FileNode> = IndexMap::new();
+      let mut globbed_directories: IndexMap<PathBuf, remexec::DirectoryNode> = IndexMap::new();
+      let mut todo_directories: IndexMap<PathBuf, Vec<RestrictedPathGlob>> = IndexMap::new();
+
+      // Populate globbed_{files,directories} by iterating over all the globs.
+      for path_glob in include.into_iter() {
+        let wildcard = match &path_glob {
+          RestrictedPathGlob::Wildcard { wildcard } => wildcard,
+          RestrictedPathGlob::DirWildcard { wildcard, .. } => wildcard,
+        };
+
+        let matching_files: Vec<PathBuf> = cur_dir_files
+          .keys()
+          .filter(|path| {
+            // NB: match just the current path component against the wildcard, but use the prefix
+            // when checking against the `exclude` patterns!
+            wildcard.matches_path(path) && !exclude.is_ignored_path(&prefix.join(path), false)
+          })
+          .cloned()
+          .collect();
+        for file_path in matching_files.into_iter() {
+          // NB: remove any matched files from `cur_dir_files`, so they are successfully matched
+          // against at most once.
+          let file_node = cur_dir_files.remove(&file_path).unwrap();
+          globbed_files.insert(file_path, file_node);
+        }
+
+        let matching_directories: Vec<PathBuf> = cur_dir_directories
+          .keys()
+          .filter(|path| {
+            // NB: match just the current path component against the wildcard, but use the prefix
+            // when checking against the `exclude` patterns!
+            wildcard.matches_path(path) && !exclude.is_ignored_path(&prefix.join(path), true)
+          })
+          .cloned()
+          .collect();
+        for directory_path in matching_directories.into_iter() {
+          // NB: do *not* remove the directory for `cur_dir_directories` after it is matched
+          // successfully once!
+          let directory_node = cur_dir_directories.get(&directory_path).unwrap();
+
+          match &path_glob {
```

Comment:
The main concerns that turned me off to that were:
1. the `scandir()` method of `VFS` expecting that accessing an arbitrary subdirectory path would be cheap to do (while `remexec::Directory` instances can only cheaply `scandir()` the top-level paths).
2. the fact that I could not see an analogue of the `/**` and `/**/*` coalescing optimization that is implemented in this PR for arbitrary `VFS` instances. One of the key reasons we can avoid an extra cache in this solution is that we do not require enumerating all of the `PathStat`s recursively contained in a Snapshot. I thought that was a fundamental difference between the approaches and was why I elected to just move things into a separate file.

Reply:
For (1), it would be possible to create a wrapper struct that impls `VFS` and supports the `scandir()`ing of arbitrary directories by an internally mutable hash map but that seems like it would require complex synchronization logic.

Reply:
And fwiw, I did actually try to `impl VFS` at first, and I ran into the issues above. I'm totally willing to reconsider if we can figure out a way to merge the two concepts. I just feel like the difference between what we want here ("do *not* recurse into subdirectories if the glob matches, for performance") and what we want from `VFS` ("we *must* recurse into subdirectories when the glob matches, to get the current filesystem state") is large enough to merit a temporary duplication of the logic here.

I 100% agree that it's not good to have this duplication and that it should be a priority to remove. I will make a followup issue to fix this if that's ok?

Reply:
I have added a TODO pointing to #9967 here. Do you think that's sufficient here?

Reply:
> the scandir() method of VFS expecting that accessing an arbitrary subdirectory path would be cheap to do (while remexec::Directory instances can only cheaply scandir() the top-level paths).

This is valid, but I think that it could be resolved pretty easily with memoization (lazy loading, effectively) of the directory structure.

> "do not recurse into subdirectories if the glob matches, for performance") and what we want from VFS ("we must recurse into subdirectories when the glob matches, to get the current filesystem state")

Ok... I see what you're saying here. I'll take another look at this tomorrow.

Reply:
> I think that it could be resolved pretty easily with memoization

I totally agree! That's what I was thinking of for https://github.com/pantsbuild/pants/pull/9779#discussion_r435626637. 

Thinking now, I think that because we can rely on the glob matching implementation to never "jump" directory components (i.e. request a path to `scandir()` that is not the immediate child of a directory that has already been scanned, a memoization approach would definitely work. Thanks!